src/System.Net.Security/src/System/Net/Security/SslState.cs

```diff
@@ -1252,6 +1237,39 @@ internal int CheckEnqueueRead(byte[] buffer, int offset, int count, AsyncProtoco
             }
         }
 
+        internal ValueTask<int> CheckEnqueueReadAsync(byte[] buffer, int offset, int count)
+        {
+            int lockState = Interlocked.CompareExchange(ref _lockReadState, LockRead, LockNone);
+
+            if (lockState != LockHandshake)
+            {
+                // Proceed, no concurrent handshake is ongoing so no need for a lock.
+                return new ValueTask<int>(CheckOldKeyDecryptedData(buffer, offset, count));
+            }
+
+            lock (this)
+            {
+                int result = CheckOldKeyDecryptedData(buffer, offset, count);
+                if (result != -1)
+                {
+                    return new ValueTask<int>(result);
+                }
+
+                // Check again under lock.
+                if (_lockReadState != LockHandshake)
+                {
+                    // The other thread has finished before we grabbed the lock.
+                    _lockReadState = LockRead;
+                    return new ValueTask<int>(-1);
+                }
+
+                _lockReadState = LockPendingRead;
+                TaskCompletionSource<int> tcs = new TaskCompletionSource<int>(new ArraySegment<byte>(buffer, offset, count), TaskCreationOptions.RunContinuationsAsynchronously);
```

Comment:
Presumably this is rare and this doesn't really matter, but the ArraySegment is going to be boxes here. Instead you could derive a type from TCS that has the state on it as fields.